epsagon/wrappers/fastapi.py

```diff
@@ -21,82 +21,230 @@
+        epsagon.trace.trace_factory.send_traces()
+    except Exception as exception:  # pylint: disable=W0703
+        print_debug('Failed to send traces: {}'.format(exception))
+        epsagon.trace.trace_factory.pop_trace(trace=trace)
+
+    if raised_err:
+        raise raised_err
+
+    return response
+
+# pylint: disable=too-many-statements
+def _wrap_handler(dependant):
+    """
+    Wraps the endppint handler.
+    """
+    original_handler = dependant.call
+    original_request_param_name = dependant.request_param_name
+    is_async = asyncio.iscoroutinefunction(original_handler)
+    dependant.request_param_name = EPSAGON_REQUEST_PARAM_NAME
+    def sync_wrapped_handler(*args, **kwargs):
+        """
+        Synchronous wrapper handler
+        """
+        request: Request = _handle_wrapper_params(
+            args, kwargs, original_request_param_name
+        )
+        if not request or not _switch_tracer_mode(is_async):
+            return original_handler(*args, **kwargs)
+        trace = None
+        should_ignore_request = True
+        try:
+            if not ignore_request('', request.url.path.lower()):
+                should_ignore_request = False
+                trace = epsagon.trace.trace_factory.get_or_create_trace()
+                trace.prepare()
+
+        except Exception as exception: # pylint: disable=W0703
+            if trace:
+                epsagon.trace.trace_factory.pop_trace(trace=trace)
+            return original_handler(*args, **kwargs)
+
+        if should_ignore_request:
+            return original_handler(*args, **kwargs)
+
+        runner = None
+        response = None
+        try:
+            runner = FastapiRunner(time.time(), request)
+            trace.set_runner(runner)
+        except Exception as exception: # pylint: disable=W0703
+            print_debug('Failed to add FastAPI runner event, skipping trace')
+            # failed to add runner event, skipping trace
+            epsagon.trace.trace_factory.pop_trace(trace=trace)
+            return original_handler(*args, **kwargs)
+        try:
+            collect_container_metadata(runner.resource['metadata'])
+        except Exception as exception: # pylint: disable=W0703
+            warnings.warn(
+                'Could not extract container metadata',
+                EpsagonWarning
+            )
+        raised_err = None
+        try:
+            response: Response = original_handler(*args, **kwargs)
+        except Exception as exception:  # pylint: disable=W0703
+            raised_err = exception
+            traceback_data = get_traceback_data_from_exception(exception)
+            trace.runner.set_exception(exception, traceback_data)
+        try:
+            loop = asyncio.new_event_loop()
+            trace.runner.update_request_body(
+                json.dumps(loop.run_until_complete(request.json()))
+            )
+
+        except json.decoder.JSONDecodeError:
+            pass
```

Comment:
maybe a debug print here as well, so if there's any weird behavior around it we'll know?

Reply:
deleted the debug print as it was spammy (it raises for every request without a body)